lib/string.js

```diff
@@ -196,6 +198,59 @@ internals.String.prototype.email = function (isEmailOptions) {
 };
 
 
+internals.String.prototype.uri = function (uriOptions) {
+    var regex = internals.uriRegex,
+        customScheme;
+
+    if (uriOptions) {
+        Hoek.assert(typeof uriOptions === 'object', 'uri options must be an object');
+
+        if (uriOptions.scheme) {
+            Hoek.assert(uriOptions.scheme instanceof RegExp || typeof uriOptions.scheme === 'string' || Array.isArray(uriOptions.scheme), 'scheme must be a RegExp, String, or Array');
+            // If someone wants to match HTTP or HTTPS for example then we need to support both RegExp and String so we don't escape their pattern unknowingly.
+            if (uriOptions.scheme instanceof RegExp) {
+                customScheme = uriOptions.scheme.source;
+            } else if (typeof uriOptions.scheme === 'string') {
+                Hoek.assert(/[a-zA-Z][a-zA-Z0-9+-\.]*/.test(uriOptions.scheme), 'scheme must be a valid scheme');
+                customScheme = Hoek.escapeRegex(uriOptions.scheme);
+            } else {
+                // With the above assertions an Array is the only thing left.
+
+                // Flatten the array into a string to be used to match the schemes.
+                customScheme = uriOptions.scheme.reduce(function(accumulator, scheme, index) {
+                    Hoek.assert(scheme instanceof RegExp || typeof scheme === 'string', 'scheme at position ' + index + ' must be a RegExp or String');
+                    // Add OR separators if a value already exists
+                    accumulator = accumulator ? accumulator + '|' : '';
+
+                    if (scheme instanceof RegExp) {
+                        return accumulator + scheme.source;
+                    } else {
+                        Hoek.assert(/[a-zA-Z][a-zA-Z0-9+-\.]*/.test(scheme), 'scheme at position ' + index + ' must be a valid scheme');
+                        return accumulator + Hoek.escapeRegex(scheme);
+                    }
+                }, '');
+            }
+        }
+    }
+
+    if (customScheme) {
+        regex = internals.createUriRegex(customScheme);
+    }
+
+    return this._test('uri', uriOptions, function (value, state, options) {
+
+        if (regex.test(value)) {
+            return null;
+        }
+
+        if (customScheme) {
+            return Errors.create('string.uriCustomScheme', { scheme: customScheme }, state, options);
```

Comment:
Can you put `value` in both errors ? Helps if people want custom messages to contain the value that failed.